pkg/storage/ingest/partition_offset_reader.go

```diff
@@ -0,0 +1,203 @@
+)
+
+var (
+	errPartitionOffsetReaderStopped = errors.New("partition offset reader is stopped")
+)
+
+// partitionOffsetReader is responsible to read the offsets of a single partition.
+//
+// If in the future we'll need to read offsets of multiple partitions at once, then we shouldn't use
+// this structure but create a new one which fetches multiple partition offsets in a single request.
+type partitionOffsetReader struct {
+	services.Service
+
+	client      *kgo.Client
+	logger      log.Logger
+	topic       string
+	partitionID int32
+
+	// nextResultWaiter is the waiter that will be notified about the result of the *next* "last produced offset"
+	// request that will be issued (not the current in-flight one, if any).
+	nextResultWaiterMx sync.RWMutex
+	nextResultWaiter   *resultWaiter[int64]
+
+	// Metrics.
+	lastProducedOffsetRequestsTotal prometheus.Counter
+	lastProducedOffsetFailuresTotal prometheus.Counter
+	lastProducedOffsetLatency       prometheus.Summary
+}
+
+func newPartitionOffsetReader(client *kgo.Client, topic string, partitionID int32, pollFrequency time.Duration, reg prometheus.Registerer, logger log.Logger) *partitionOffsetReader {
+	p := &partitionOffsetReader{
+		client:           client,
+		topic:            topic,
+		partitionID:      partitionID,
+		logger:           logger, // Do not wrap with partition ID because it's already done by the caller.
+		nextResultWaiter: newResultWaiter[int64](),
+
+		lastProducedOffsetRequestsTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name:        "cortex_ingest_storage_reader_last_produced_offset_requests_total",
+			Help:        "Total number of requests issued to get the last produced offset.",
+			ConstLabels: prometheus.Labels{"partition": strconv.Itoa(int(partitionID))},
+		}),
+		lastProducedOffsetFailuresTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name:        "cortex_ingest_storage_reader_last_produced_offset_failures_total",
+			Help:        "Total number of failed requests to get the last produced offset.",
+			ConstLabels: prometheus.Labels{"partition": strconv.Itoa(int(partitionID))},
+		}),
+		lastProducedOffsetLatency: promauto.With(reg).NewSummary(prometheus.SummaryOpts{
+			Name:        "cortex_ingest_storage_reader_last_produced_offset_request_duration_seconds",
+			Help:        "The duration of requests to fetch the last produced offset of a given partition.",
+			ConstLabels: prometheus.Labels{"partition": strconv.Itoa(int(partitionID))},
+			Objectives:  latencySummaryObjectives,
+			MaxAge:      time.Minute,
+			AgeBuckets:  10,
+		}),
+	}
+
+	p.Service = services.NewTimerService(pollFrequency, nil, p.onPollInterval, p.stopping)
+
+	return p
+}
+
+func (p *partitionOffsetReader) onPollInterval(ctx context.Context) error {
+	// The following call blocks until the last produced offset has been fetched from Kafka. If fetching
+	// the offset takes longer than the poll interval, than we'll poll less frequently than configured.
```

Comment:
Yes, we could, but I think it's something we can defer an a later stage optimization if we see being an issue in practice. The reason why a simple edging (using the same Kafka client) may not help is because requests are serialized, so the response from request N+1 will only come after the response to request N. I think that do a real edging we'll have to use different Kafka clients.

Reply:
> The reason why a simple edging (using the same Kafka client) may not help is because requests are serialized, so the response from request N+1 will only come after the response to request N

At least, this is what I understood reading the client code, cause I also had the same thought about edging. Better investigation should be done before implementing it.

Reply:
good point. I had forgotten about this. This makes things more complicated